docs/bricks_design.rst

```diff
@@ -0,0 +1,602 @@
+designed. We will start with the simplest approach possible and show
+step by step what benefits more advanced methods bring. The final goal is
+to explain the ideas at the core of the respective part of Blocks. 
+In Blocks the components assisting to build the computation
+graph are called "bricks". Bricks are a unified representation for
+components of all levels: basic ones, like activation functions, and
+complex ones, like attention mechanism and a generic sequence generator.
+Reading this document should answer many questions that people
+often ask about bricks, such as "why so many decorators", "why is not my
+brick initialized in the constructor" and others, which are asked not so
+often, but still daunt people' minds.
+
+.. _PyLearn2: https://github.com/lisa-lab/pylearn2
+.. _Groundhog: https://github.com/lisa-groundhog/GroundHog
+.. _Lasagne: https://github.com/Lasagne/Lasagne
+.. _Keras: https://github.com/fchollet/keras
+
+Design for basic components
+---------------------------
+
+The toy task
+~~~~~~~~~~~~
+
+To make our discussion less abstract, let us consider an example. D,
+a novice in deep learning, has learnt about fully-connected feed-forward
+and recurrent networks (FFN and RNN respectively). He plans to use them
+in a number of projects, so he decides to start with coding a reference
+implementation first. Few hours of hard work and here it is:
+
+    >>> # D's FFN, D2L2v0.1
+    >>> inputs = tensor.matrix('x')
+    >>> inputs.tag.test_value = numpy.random.uniform(size=(10, 754))
+    >>> dims = [754, 100, 100, 10]
+    >>> activations = [tensor.tanh, tensor.tanh, tensor.nnet.softmax]
+    >>> weight_matrices = []
+    >>> bias_vectors = []
+    >>> for i in range(len(dims) - 1):
+    >>>     weight_matrices.append(theano.shared(numpy.random.uniform(
+    >>>         size=(dims[i], dims[i + 1]))))
+    >>>     bias_vectors.append(theano.shared(numpy.zeros(dims[i + 1])))
+    >>> outputs = inputs
+    >>> for dim, activation, weight_matrix, bias_vector in zip(
+    >>>         dims[1:], activations, weight_matrices, bias_vectors):
+    >>>     outputs = activation(outputs.dot(weight_matrix) + bias_vector)
+    >>> 
+    >>> # D's RNN, D2L2v0.1
+    >>> inputs = tensor.tensor3('x')
+    >>> inputs.tag.test_value = numpy.random.uniform(size=(5, 10, 100))
+    >>> dim = 100
+    >>> weight_matrix = theano.shared(numpy.random.uniform(size=(dim, dim)))
+    >>> bias_vector = theano.shared(numpy.random.uniform(size=dim))
+    >>> outputs, _ = theano.scan(
+    >>>     lambda states, _inputs, _weight_matrix: states.dot(_weight_matrix) + _inputs,
+    >>>     sequences=[inputs], non_sequences=[weight_matrix],
+    >>>     outputs_info=tensor.zeros_like(inputs[0]))
+
+The two code snippets above has formed D's brand-new DL toolkit D2L2
+(D Deep Learning Library). He picks 0.1 as the initial version
+identifies. Hooray, an new project has been born!
+
+Well, why not just copy-paste?!
```

Comment:
I agree with Dima, maybe the software engineering related part could be moved into a separate chapter, so that who is familiar with it can just skip it.